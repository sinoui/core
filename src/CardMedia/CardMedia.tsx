import styled, { css } from 'styled-components';
import classNames from 'classnames';

interface Props {
  /**
   * 图片链接
   */
  imageUrl: string;
  /**
   * 在水平方向上图片占满卡片宽度。采用 16:9 的比例。
   */
  wide?: boolean;
  /**
   * 图片呈现为一个正方形小块。
   */
  square?: boolean;
  /**
   * 图片宽度
   */
  width?: number;
}

const squareStyle = css`
  &::before {
    margin-top: 100%;
  }
`;

const wideStyle = css`
  &::before {
    margin-top: 56.25%;
  }
`;

const CardMediaClassName = 'sinoui-card__media';

const CardMediaWrapper = styled.div.attrs(({ wide, square }: Props) => ({
  className: classNames(CardMediaClassName, {
    [`${CardMediaClassName}--16-9`]: wide,
    [`${CardMediaClassName}--square`]: square,
  }),
}))<Props>`
  background-image: ${({ imageUrl }) => `url(${imageUrl})`};
  background-repeat: no-repeat;
  background-position: 50%;
  background-size: cover;

  position: relative;
  overflow: hidden;

  &::before {
    content: '';
    display: block;
  }

  ${({ square }) => square && squareStyle}
  ${({ wide }) => wide && wideStyle}

  ${({ width }) =>
    width &&
    `
    width: ${width}px;
  `}
  box-sizing: border-box;
`;

if (process.env.NODE_ENV === 'development') {
  CardMediaWrapper.displayName = 'CardMedia';
}

export default CardMediaWrapper;
