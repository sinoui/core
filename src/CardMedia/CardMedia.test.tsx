/**
 * @jest-environment jsdom
 */
import React from 'react';
import renderer from 'react-test-renderer';
import { render, cleanup } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { defaultTheme } from '@sinoui/theme';
import '@testing-library/jest-dom';
import CardMedia from './CardMedia';

const mediaImgUrl =
  'https://material-components.github.io/material-components-web-catalog/static/media/photos/3x2/2.jpg';

describe('CardMedia组件单元测试', () => {
  afterEach(cleanup);

  it('CardMedia设置wide属性', () => {
    const { container } = render(
      <ThemeProvider theme={defaultTheme}>
        <CardMedia imageUrl={mediaImgUrl} wide />
      </ThemeProvider>,
    );
    expect(container.firstChild).toHaveClass('sinoui-card__media--16-9');
  });

  it('CardMedia设置square属性', () => {
    const { container } = render(
      <ThemeProvider theme={defaultTheme}>
        <CardMedia imageUrl={mediaImgUrl} square />
      </ThemeProvider>,
    );
    expect(container.firstChild).toHaveClass('sinoui-card__media--square');
  });
});

describe('快照测试', () => {
  it('CardMedia设置wide属性', () => {
    const tree = renderer
      .create(
        <ThemeProvider theme={defaultTheme}>
          <CardMedia imageUrl={mediaImgUrl} wide />
        </ThemeProvider>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('CardMedia设置square属性', () => {
    const tree = renderer
      .create(
        <ThemeProvider theme={defaultTheme}>
          <CardMedia imageUrl={mediaImgUrl} square />
        </ThemeProvider>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
