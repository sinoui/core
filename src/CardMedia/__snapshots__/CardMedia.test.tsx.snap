// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`快照测试 CardMedia设置square属性 1`] = `
<div
  className="sc-AxjAm dyhTnR sinoui-card__media sinoui-card__media--square"
/>
`;

exports[`快照测试 CardMedia设置wide属性 1`] = `
<div
  className="sc-AxjAm dQNili sinoui-card__media sinoui-card__media--16-9"
/>
`;
